File: frontend/js/manage_logging.js
/**
 * @file manage_logging.js
 * @description Verwaltungsmodul für Logging in Fotobox2
 * @module manage_logging
 */

import { apiPost, apiGet } from './manage_api.js';
import { Result } from './utils.js';

// API-Endpunkte
const API = {
    LOG: '/api/logging/log',
    CONFIG: '/api/logging/config',
    STATUS: '/api/logging/status',
    HISTORY: '/api/logging/history'
};

/**
 * @typedef {'debug'|'info'|'warn'|'error'} LogLevel
 */

/**
 * @typedef {Object} LogEntry
 * @property {LogLevel} level - Log-Level
 * @property {string} message - Nachricht
 * @property {Object} [context] - Zusätzlicher Kontext
 * @property {string} timestamp - Zeitstempel
 * @property {'frontend'|'backend'} source - Quelle des Logs
 * @property {string} [component] - Betroffene Komponente
 */

/**
 * @typedef {Object} LogConfig
 * @property {LogLevel} minLevel - Minimales Log-Level
 * @property {boolean} enableConsole - Konsolen-Logging aktiv
 * @property {boolean} enableServer - Server-Logging aktiv
 * @property {number} maxQueueSize - Maximale Queue-Größe
 * @property {boolean} logStackTraces - Stack-Traces loggen
 */

// Konfiguration
const LOG_LEVELS = {
    debug: 0,
    info: 1,
    warn: 2,
    error: 3
};

// Standard-Konfiguration
let config = {
    minLevel: 'info',
    enableConsole: true,
    enableServer: true,
    maxQueueSize: 100,
    logStackTraces: true
};

// Queue für offline/fehlerhafte Logs
let logQueue = [];

/**
 * Aktualisiert die Logging-Konfiguration
 * @param {Partial<LogConfig>} newConfig - Neue Konfigurationsoptionen
 */
export async function configure(newConfig) {
    try {
        const response = await apiPost(API.CONFIG, newConfig);
        if (response.success) {
            config = { ...config, ...newConfig };
            debug('Logging-Konfiguration aktualisiert');
            return Result.ok(config);
        }
        return Result.fail(response.error);
    } catch (err) {
        console.error('Fehler bei Logging-Konfiguration:', err);
        return Result.fail(err.message);
    }
}

/**
 * Sendet einen Log an den Server
 * @param {LogEntry} entry - Der zu sendende Log-Eintrag
 * @returns {Promise<Result>} Erfolg oder Fehler
 */
async function sendLogToServer(entry) {
    if (!config.enableServer) return Result.ok(null);
    
    try {
        const response = await apiPost(API.LOG, entry);
        if (response.success) {
            return Result.ok(null);
        }
        // Bei Fehler in Queue speichern
        if (logQueue.length < config.maxQueueSize) {
            logQueue.push(entry);
        }
        return Result.fail(response.error);
    } catch (err) {
        if (logQueue.length < config.maxQueueSize) {
            logQueue.push(entry);
        }
        return Result.fail('Server nicht erreichbar');
    }
}

/**
 * Versucht, aufgestaute Logs aus der Queue zu senden
 */
export async function processLogQueue() {
    if (logQueue.length === 0) return;
    
    try {
        // Wir senden maximal 10 Logs auf einmal
        const logsToSend = logQueue.splice(0, 10);
        await apiPost('/api/logs/batch', { logs: logsToSend });
        
        // Wenn erfolgreich, versuchen wir weitere Logs zu senden
        if (logQueue.length > 0) {
            setTimeout(processLogQueue, 1000);
        }
    } catch (err) {
        // Bei Fehler warten wir länger, bevor wir es erneut versuchen
        setTimeout(processLogQueue, 5000);
    }
}

/**
 * Loggt eine Nachricht
 * @param {LogLevel} level - Das Log-Level
 * @param {string} message - Die zu loggende Nachricht
 * @param {Object} [context] - Optionaler Kontext für die Nachricht
 */
export function log(level, message, context = null) {
    if (LOG_LEVELS[level] === undefined) {
        console.warn(`Unbekanntes Log-Level: ${level}`);
        return;
    }
    
    if (LOG_LEVELS[level] >= LOG_LEVELS[config.minLevel]) {
        if (config.enableConsole) {
            console.log(`[${level.toUpperCase()}] ${message}`, context ? context : '');
        }
        sendLogToServer({ level, message, context, timestamp: new Date().toISOString(), source: 'frontend' });
    }
}

/**
 * Loggt eine Informationsnachricht
 * @param {string} message - Die zu loggende Nachricht
 * @param {Object} [context] - Optionaler Kontext für die Nachricht
 */
export function info(message, context = null) {
    log('info', message, context);
}

/**
 * Loggt eine Fehlernachricht
 * @param {string} message - Die zu loggende Fehlernachricht
 * @param {Error|Object} [err] - Der Fehler oder zusätzlicher Kontext
 */
export function error(message, err = null) {
    if (LOG_LEVELS.error >= LOG_LEVELS[config.minLevel]) {
        if (config.enableConsole) {
            console.error(`[ERROR] ${message}`, err ? err : '');
        }
        
        let context = null;
        if (err instanceof Error) {
            context = {
                name: err.name,
                message: err.message,
                stack: err.stack
            };
        } else if (err) {
            context = err;
        }
        
        sendLogToServer({ level: 'error', message, context, timestamp: new Date().toISOString(), source: 'frontend' });
    }
}

/**
 * Loggt eine Warnungsnachricht
 * @param {string} message - Die zu loggende Warnungsnachricht
 * @param {Object} [context] - Optionaler Kontext für die Nachricht
 */
export function warn(message, context = null) {
    log('warn', message, context);
}

/**
 * Loggt eine Debug-Nachricht (nur im Entwicklungsmodus)
 * @param {string} message - Die zu loggende Debug-Nachricht
 * @param {Object} [context] - Optionaler Kontext für die Nachricht
 */
export function debug(message, context = null) {
    log('debug', message, context);
}

/**
 * Ruft Logs vom Server ab
 * @param {Object} options - Optionen für die Abfrage
 * @param {string} [options.level] - Filter nach Log-Level
 * @param {number} [options.limit=100] - Maximale Anzahl abzurufender Logs
 * @param {number} [options.offset=0] - Offset für Paginierung
 * @param {string} [options.startDate] - Filter - Logs nach diesem Datum (ISO-Format)
 * @param {string} [options.endDate] - Filter - Logs vor diesem Datum (ISO-Format)
 * @param {string} [options.source] - Filter nach Log-Quelle
 * @returns {Promise<Array>} - Liste von Logeinträgen
 */
export async function getLogs(options = {}) {
    try {
        const response = await apiPost('/api/logs', options);
        return response.logs || [];
    } catch (err) {
        error('Fehler beim Abrufen der Logs', err);
        return [];
    }
}

/**
 * Löscht Logs
 * @param {Object} options - Optionen für das Löschen (z.B. Zeitraum)
 * @returns {Promise<boolean>} true wenn erfolgreich
 */
export async function clearLogs(options = {}) {
    try {
        const response = await apiPost('/api/logs/clear', options);
        return response.success || false;
    } catch (err) {
        error('Fehler beim Löschen der Logs', err);
        return false;
    }
}

// Initialisierung und regelmäßige Verarbeitung der Log-Queue
setTimeout(processLogQueue, 5000);

// Initialisierung des Logging-Moduls
log('info', 'Logging-Modul initialisiert');

File: backend/manage_logging.py
#!/usr/bin/env python3
"""
@file manage_logging.py
@description Verwaltungsmodul für Logging in Fotobox2
@module manage_logging
"""

import os
import logging
import json
import sqlite3
from datetime import datetime
import traceback

# Initialisiere Basis-Logging für Bootstrapping-Phase
bootstrap_logger = logging.getLogger('bootstrap')
console_handler = logging.StreamHandler()
console_handler.setFormatter(logging.Formatter('[%(levelname)s] %(message)s'))
bootstrap_logger.addHandler(console_handler)
bootstrap_logger.setLevel(logging.INFO)

# Pfadkonfiguration über das zentrale Verzeichnismanagement
try:
    from manage_folders import get_log_dir, get_data_dir
    LOG_DIR = get_log_dir()
    DB_DIR = get_data_dir()
    
    # Stelle sicher, dass die Verzeichnisse existieren
    os.makedirs(LOG_DIR, exist_ok=True)
    os.makedirs(DB_DIR, exist_ok=True)
    
    bootstrap_logger.info(f"Log-Verzeichnis initialisiert: {LOG_DIR}")
    bootstrap_logger.info(f"Datenbank-Verzeichnis initialisiert: {DB_DIR}")
    
except ImportError as e:
    bootstrap_logger.error(f"Fehler beim Import von manage_folders: {e}")
    # Fallback falls manage_folders nicht verfügbar ist
    LOG_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'log'))
    DB_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'data'))
    
    try:
        os.makedirs(LOG_DIR, exist_ok=True)
        os.makedirs(DB_DIR, exist_ok=True)
        bootstrap_logger.info("Fallback-Verzeichnisse erstellt")
    except Exception as e:
        bootstrap_logger.error(f"Kritischer Fehler bei Verzeichniserstellung: {e}")
        raise

DB_PATH = os.path.join(DB_DIR, 'fotobox_logs.db')
LOG_FILE = os.path.join(LOG_DIR, f"{datetime.now().strftime('%Y-%m-%d')}_fotobox.log")
DEBUG_LOG_FILE = os.path.join(LOG_DIR, f"{datetime.now().strftime('%Y-%m-%d')}_fotobox_debug.log")

# Logging-Level konfigurieren - kann aus einer Konfigurationsdatei geladen werden
DEFAULT_LOG_LEVEL = logging.INFO

try:
    # Setup für Dateilogging
    file_handler = logging.FileHandler(LOG_FILE)
    file_handler.setFormatter(logging.Formatter('%(asctime)s [%(levelname)s] %(message)s'))

    # Setup für Debug-Logging
    debug_handler = logging.FileHandler(DEBUG_LOG_FILE)
    debug_handler.setFormatter(logging.Formatter('%(asctime)s [%(levelname)s] %(message)s - %(pathname)s:%(lineno)d'))
    debug_handler.setLevel(logging.DEBUG)

    # Root Logger konfigurieren
    logger = logging.getLogger('fotobox')
    logger.setLevel(DEFAULT_LOG_LEVEL)
    logger.addHandler(file_handler)
    logger.addHandler(debug_handler)
    logger.addHandler(console_handler)
    
    bootstrap_logger.info("Logging-System erfolgreich initialisiert")
    
except Exception as e:
    bootstrap_logger.error(f"Fehler bei der Logger-Initialisierung: {e}")
    raise

def _init_db():
    """
    Initialisiert die Datenbank für die Log-Speicherung
    """
    try:
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        cursor.execute('''
        CREATE TABLE IF NOT EXISTS logs (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            timestamp TEXT NOT NULL,
            level TEXT NOT NULL,
            message TEXT NOT NULL,
            context TEXT,
            source TEXT,
            user_id TEXT
        )
        ''')
        conn.commit()
        conn.close()
    except Exception as e:
        # Fallback zu Dateilogging, wenn DB nicht verfügbar ist
        logger.error(f"Fehler bei DB-Initialisierung: {e}")

def _store_log_in_db(level, message, context=None, source=None, user_id=None):
    """
    Speichert einen Logeintrag in der Datenbank
    
    Args:
        level: Log-Level (INFO, WARNING, ERROR, etc.)
        message: Die Lognachricht
        context: Zusätzlicher Kontext als Dict
        source: Quelle des Logs (Funktion, Modul, etc.)
        user_id: Optionale Benutzer-ID
    """
    try:
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        timestamp = datetime.now().isoformat()
        context_json = json.dumps(context) if context else None
        cursor.execute(
            "INSERT INTO logs (timestamp, level, message, context, source, user_id) VALUES (?, ?, ?, ?, ?, ?)",
            (timestamp, level, message, context_json, source, user_id)
        )
        conn.commit()
        conn.close()
    except Exception as e:
        # Falls ein Fehler beim DB-Speichern auftritt, loggen wir in die Datei
        logger.error(f"Fehler beim Speichern des Logs in DB: {e}")

def log(message, context=None, source=None, user_id=None):
    """
    Loggt eine Informationsnachricht
    
    Args:
        message: Die zu loggende Nachricht
        context: Optionaler Kontext als Dict
        source: Quelle des Logs (Funktion, Modul, etc.)
        user_id: Optionale Benutzer-ID
    """
    logger.info(message)
    _store_log_in_db("INFO", message, context, source, user_id)

def error(message, exception=None, context=None, source=None, user_id=None):
    """
    Loggt eine Fehlernachricht
    
    Args:
        message: Die zu loggende Fehlernachricht
        exception: Optionale Exception
        context: Optionaler Kontext als Dict
        source: Quelle des Logs (Funktion, Modul, etc.)
        user_id: Optionale Benutzer-ID
    """
    if exception:
        error_details = {
            "exception_type": type(exception).__name__,
            "exception_message": str(exception),
            "traceback": traceback.format_exc()
        }
        if context:
            context.update(error_details)
        else:
            context = error_details
        logger.error(f"{message}: {exception}")
    else:
        logger.error(message)
    
    _store_log_in_db("ERROR", message, context, source, user_id)

def warn(message, context=None, source=None, user_id=None):
    """
    Loggt eine Warnungsnachricht
    
    Args:
        message: Die zu loggende Warnungsnachricht
        context: Optionaler Kontext als Dict
        source: Quelle des Logs (Funktion, Modul, etc.)
        user_id: Optionale Benutzer-ID
    """
    logger.warning(message)
    _store_log_in_db("WARNING", message, context, source, user_id)

def debug(message, context=None, source=None, user_id=None):
    """
    Loggt eine Debug-Nachricht (nur im Debug-Modus)
    
    Args:
        message: Die zu loggende Debug-Nachricht
        context: Optionaler Kontext als Dict
        source: Quelle des Logs (Funktion, Modul, etc.)
        user_id: Optionale Benutzer-ID
    """
    logger.debug(message)
    # Debug-Logs werden optional auch in DB gespeichert
    if logger.level <= logging.DEBUG:
        _store_log_in_db("DEBUG", message, context, source, user_id)

def get_logs(level=None, limit=100, offset=0, start_date=None, end_date=None, source=None):
    """
    Ruft Logs aus der Datenbank ab
    
    Args:
        level: Optional - Filter nach Log-Level
        limit: Maximale Anzahl zurückzugebender Logs
        offset: Offset für Paginierung
        start_date: Filter - Logs nach diesem Datum (ISO-Format)
        end_date: Filter - Logs vor diesem Datum (ISO-Format)
        source: Filter nach Log-Quelle
        
    Returns:
        Liste von Logeinträgen
    """
    try:
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        
        query = "SELECT id, timestamp, level, message, context, source, user_id FROM logs"
        conditions = []
        params = []
        
        if level:
            conditions.append("level = ?")
            params.append(level)
            
        if start_date:
            conditions.append("timestamp >= ?")
            params.append(start_date)
            
        if end_date:
            conditions.append("timestamp <= ?")
            params.append(end_date)
            
        if source:
            conditions.append("source = ?")
            params.append(source)
            
        if conditions:
            query += " WHERE " + " AND ".join(conditions)
            
        query += " ORDER BY timestamp DESC LIMIT ? OFFSET ?"
        params.extend([limit, offset])
        
        cursor.execute(query, params)
        rows = cursor.fetchall()
        
        logs = []
        for row in rows:
            log_entry = {
                "id": row[0],
                "timestamp": row[1],
                "level": row[2],
                "message": row[3],
                "context": json.loads(row[4]) if row[4] else None,
                "source": row[5],
                "user_id": row[6]
            }
            logs.append(log_entry)
            
        conn.close()
        return logs
    except Exception as e:
        logger.error(f"Fehler beim Abrufen der Logs: {e}")
        return []

def clear_logs(older_than=None):
    """
    Löscht Logs aus der Datenbank
    
    Args:
        older_than: Optional - Nur Logs älter als dieses Datum (ISO-Format) löschen
        
    Returns:
        Anzahl gelöschter Logeinträge
    """
    try:
        conn = sqlite3.connect(DB_PATH)
        cursor = conn.cursor()
        
        if older_than:
            cursor.execute("DELETE FROM logs WHERE timestamp < ?", (older_than,))
        else:
            cursor.execute("DELETE FROM logs")
            
        deleted_count = cursor.rowcount
        conn.commit()
        conn.close()
        
        logger.info(f"{deleted_count} Logeinträge wurden gelöscht")
        return deleted_count
    except Exception as e:
        logger.error(f"Fehler beim Löschen der Logs: {e}")
        return 0

def set_log_level(level):
    """
    Setzt das Log-Level
    
    Args:
        level: Das zu setzende Log-Level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        
    Returns:
        Boolean, ob das Log-Level erfolgreich gesetzt wurde
    """
    try:
        log_level = getattr(logging, level.upper())
        logger.setLevel(log_level)
        logger.info(f"Log-Level auf {level.upper()} gesetzt")
        return True
    except (AttributeError, TypeError):
        logger.error(f"Ungültiges Log-Level: {level}")
        return False

# Initialisiere die Datenbank beim Import des Moduls
_init_db()
